src/bp/nlu-core/intents/oos-intent-classfier.ts

```diff
@@ -0,0 +1,245 @@
+import { MLToolkit } from 'botpress/sdk'
+import _ from 'lodash'
+import { isPOSAvailable } from 'nlu-core/language/pos-tagger'
+import { getStopWordsForLang } from 'nlu-core/language/stopWords'
+import {
+  featurizeInScopeUtterances,
+  featurizeOOSUtterances,
+  getUtteranceFeatures
+} from 'nlu-core/out-of-scope-featurizer'
+import { SMALL_TFIDF } from 'nlu-core/tools/tfidf'
+import { isSpace, SPACE } from 'nlu-core/tools/token-utils'
+import { Intent, Tools } from 'nlu-core/typings'
+import Utterance, { buildUtteranceBatch } from 'nlu-core/utterance/utterance'
+
+import { BaseIntentClassifier, MIN_NB_UTTERANCES } from './base-intent-classifier'
+import { IntentTrainInput, NoneableIntentClassifier, NoneableIntentPredictions } from './intent-classifier'
+
+interface TrainInput extends IntentTrainInput {
+  allUtterances: Utterance[]
+}
+
+interface Model {
+  trainingVocab: string[]
+  baseIntentClfModel: string
+  oosSvmModel: string | undefined
+}
+
+interface Predictors {
+  baseIntentClf: BaseIntentClassifier
+  oosSvm: MLToolkit.SVM.Predictor | undefined
+  trainingVocab: string[]
+}
+
+export const NONE_INTENT = 'none'
+const NONE_UTTERANCES_BOUNDS = {
+  MIN: 20,
+  MAX: 200
+}
+
+export class OOSIntentClassifier implements NoneableIntentClassifier {
+  private model: Model | undefined
+  private predictors: Predictors | undefined
+
+  constructor(private tools: Tools) {}
+
+  public async train(trainInput: TrainInput, progress: (p: number) => void): Promise<void> {
+    const { languageCode, allUtterances } = trainInput
+    const noneIntent = await this.makeNoneIntent(allUtterances, languageCode)
+
+    const [ooScopeModel, inScopeModel] = await Promise.all([
+      this._trainOOScopeSvm(trainInput, noneIntent, (p: number) => progress(p / 2)),
```

Comment:
I've broken the progress logic right here but I have already repaired it in another branch locally.